File: bestia-zoneserver/src/main/java/net/bestia/zoneserver/entity/components/InventoryComponent.java
package net.bestia.zoneserver.entity.components;

import java.util.ArrayList;
import java.util.List;
import java.util.Optional;

import net.bestia.model.domain.Item;

/**
 * Entities having this trait can be loaded with a certain amount of items into
 * their local inventory and use this information to micro manage their own
 * inventory.
 * 
 * @author Thomas Felix
 *
 */
public class InventoryComponent extends Component {

	private static class ItemCount {
		public int amount;
		public Item item;

		public ItemCount(Item item, int amount) {

			this.amount = amount;
			this.item = item;
		}
	}

	private static final long serialVersionUID = 1L;

	private float maxWeight;
	private int maxItems;
	private List<ItemCount> items = new ArrayList<>();

	public InventoryComponent(long id) {
		super(id);
		// no op.
	}

	/**
	 * Gives the maximum carryable weight of this entity. It is pos. infinite if
	 * its unlimited.
	 * 
	 * @return The maximum item weight.
	 */
	public float getMaxWeight() {
		return maxWeight;
	}

	/**
	 * The current item weight in kg. The item weight per item is saved as 0.1kg
	 * per unit.
	 * 
	 * @return Current item weight.
	 */
	public float getWeight() {
		return items.stream().mapToInt(x -> x.amount * x.item.getWeight()).sum() / 10f;
	}

	/**
	 * The maximum number of items for this unit. Or -1 of its unlimited.
	 * 
	 * @return The maximum item count. -1 if unlimited.
	 */
	public int getMaxItemCount() {
		return maxItems;
	}

	/**
	 * The current number of item (slots) stored in this entity.
	 * 
	 * @return Current number of item slots stored in this entity.
	 */
	public int getItemCount() {
		return items.size();
	}

	/**
	 * Adds the new item to the inventory. But this will work only if the number
	 * of item slots are not exceeded and the total amount of weight does not be
	 * bigger as the maximum amount of weight.
	 * 
	 * @param item
	 *            The item to be added.
	 * @param amount
	 *            The amount of the item to be added to the inventory.
	 * @return TRUE if the item coult be added to the inventory. FALSE
	 *         otherwise.
	 */
	public boolean addItem(Item item, int amount) {
		// Check count.
		if (getItemCount() + 1 > getMaxItemCount()) {
			return false;
		}

		// Check weight.
		if (item.getWeight() / .1f + getWeight() > getMaxWeight()) {
			return false;
		}

		items.add(new ItemCount(item, amount));
		return true;
	}

	/**
	 * Removes the item from the inventory. It returns true only if the item
	 * could be successfully be removed.
	 * 
	 * @param item
	 * @param amount
	 * @return
	 */
	public boolean removeItem(Item item, int amount) {
		Optional<ItemCount> inventoryItem = items.stream().filter(x -> x.item.getId() == item.getId()).findFirst();
		if (inventoryItem.isPresent()) {
			
			ItemCount ic = inventoryItem.get();
			
			if(ic.amount > amount) {
				ic.amount -= amount;
				return true;
			} else if(ic.amount == amount) {
				items.remove(ic);
				return true;
			} else {
				return false;
			}
		} else {
			return false;
		}
	}
}

File: bestia-zoneserver/src/main/kotlin/net/bestia/zoneserver/entity/component/InventoryComponent.kt
package net.bestia.zoneserver.entity.component

import net.bestia.model.item.PlayerItemId

data class InventoryItem(
    val amount: Int,
    val weight: Int,
    val playerItemId: PlayerItemId
) {
  val totalWeight: Int
    get() = amount * weight
}

/**
 * Entities having this trait can be loaded with a certain amount of items into
 * their local inventory and use this information to micro manage their own
 * inventory.
 *
 * @author Thomas Felix
 */
data class InventoryComponent(
    override val entityId: Long,

    /**
     * Gives the maximum carryable weight of this entity. It is pos. infinite if
     * its unlimited. Its unit is measured in 0.1 increments of a kg.
     *
     * @return The maximum item weight.
     */
    val maxWeight: Int = 100,

    /**
     * The maximum number of items for this unit. Or -1 of its unlimited.
     *
     * @return The maximum item count. -1 if unlimited.
     */
    val maxItemCount: Int = 100,

    val items: List<InventoryItem> = emptyList()
) : Component {

  init {
    require(totalWeight <= maxWeight) { "Current weight ($totalWeight) is bigger then max weight ($maxWeight)" }
    require(items.size <= maxItemCount) { "Item slots used (${items.size} is bigger then max item slots ($maxItemCount)" }
  }

  /**
   * The current total item weight in units of 0.1 kg.
   *
   * @return Current item weight.
   */
  val totalWeight: Int
    get() = items.sumBy { it.totalWeight }

  companion object {
    /**
     * Constant for meaning that the item count is basically not limited.
     */
    const val UNLIMITED_ITEMS = -1
  }
}
